lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm_2019/ashrae_90_1_prm_2019.SpaceType.rb

```diff
@@ -0,0 +1,396 @@
+      # Assign data
+      lighting_per_length = sub_space_type_properties['w/ft'].to_f
+      sum_lighting_per_length += lighting_per_length
+    end
+    return space_type.spaces.size > 1 && sum_lighting_per_length > 0
+  end
+
+  # Calculate the lighting power density per area based on user data (space_based)
+  # The function will calculate the LPD based on the space type (STRING)
+  # It considers lighting per area, lighting per length as well as occupancy factors in the database.
+  # @param user_data [Hash] user data from the user csv
+  # @param space [OpenStudio::Model::Space]
+  def calculate_lpd_from_userdata(user_data, space)
+    num_std_ltg_types = user_data['num_std_ltg_types'].to_i
+    space_lighting_per_area = 0.0
+    occupancy_control_credit_sum = 0.0
+    frac_sum = 0.0 # prevent the total fraction over 1.0
+    std_ltg_index = 0 # loop index
+    # Loop through standard lighting type in a space
+    while std_ltg_index < num_std_ltg_types && frac_sum <= 1.0
+      # Retrieve data from user_data
+      type_key = 'std_ltg_type%02d' % (std_ltg_index + 1)
+      frac_key = 'std_ltg_type_frac%02d' % (std_ltg_index + 1)
+      sub_space_type = user_data[type_key]
+      sub_space_type_frac = user_data[frac_key].to_f
+      # Adjust while loop condition factors
+      frac_sum += sub_space_type_frac
+      std_ltg_index += 1
+      # get interior lighting data
+      sub_space_type_properties = interior_lighting_get_prm_data(sub_space_type)
+      # Assign data
+      lights_have_info = false
+      lighting_per_area = sub_space_type_properties['w/ft^2'].to_f
+      lighting_per_length = sub_space_type_properties['w/ft'].to_f
+      lights_have_info = true unless lighting_per_area.zero? && lighting_per_length.zero?
+      manon_or_partauto = sub_space_type_properties['manon_or_partauto'].to_i
+      # the lighting power density of this space area
+      user_space_type_lighting_area = 0.0
+      # Adjust the occupancy control sensor reduction factor from dataset
+      occ_control_reduction_factor = 0.0
+      if manon_or_partauto == 1
+        occ_control_reduction_factor = sub_space_type_properties['occup_sensor_savings'].to_f
+      else
+        occ_control_reduction_factor = sub_space_type_properties['occup_sensor_auto_on_svgs'].to_f
+      end
+
+      if lights_have_info
+        # Space height
+        space_volume = space.volume
+        space_area = space.floorArea
+        space_height = OpenStudio.convert(space_volume / space_area, 'm', 'ft').get
+        # calculate and add new lpd values
+        user_space_type_lighting_area = (lighting_per_length * space_height +
+          lighting_per_area) * sub_space_type_frac
+        space_lighting_per_area += user_space_type_lighting_area
+      end
+      # Now calculate the occupancy control credit factor (weighted by frac_lpd)
+      occupancy_control_credit_sum += occ_control_reduction_factor * user_space_type_lighting_area * sub_space_type_frac
```

Comment:
I think this also needs to be weighted by space_lighting_per_area.

Reply:
The `space_lighting_per_area` is included at line [390](https://github.com/NREL/openstudio-standards/blob/set_lpd_from_userdata/lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm_2019/ashrae_90_1_prm_2019.SpaceType.rb#L393)

Reply:
I'm a bit confused about lights_have_info. Should the occupance_control_credit_sum be inside the "if lights_have_info" block? Otherwise, user_space_type_lighting_area might not have the correct value?
Also, should the name be changed from user_space_type_lighting_area to user_space_type_lighting_per_area?

Reply:
Also, maybe we should give an error if frac_sum would exceed 1.0.

Reply:
> I'm a bit confused about lights_have_info. Should the occupance_control_credit_sum be inside the "if lights_have_info" block? Otherwise, user_space_type_lighting_area might not have the correct value? Also, should the name be changed from user_space_type_lighting_area to user_space_type_lighting_per_area?

The `lights_have_info` is copied from legacy code that checks if the data has `lighting_per_area` or `lighting_per_length`. This particularly [block](https://github.com/NREL/openstudio-standards/blob/set_lpd_from_userdata/lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm_2019/ashrae_90_1_prm_2019.SpaceType.rb#L394-L403) was created to contain the calculation of `user_type_lighting_per_area` from `lighting_per_area` and `lighting_per_length`. The calculation of `occupancy_control_credit_sum` is left outside the block but it wont be affected by the logic inside the block because we have initialized the parameters [here](https://github.com/NREL/openstudio-standards/blob/set_lpd_from_userdata/lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm_2019/ashrae_90_1_prm_2019.SpaceType.rb#L385).

However, what I think may make sense is to move this [part](https://github.com/NREL/openstudio-standards/blob/set_lpd_from_userdata/lib/openstudio-standards/standards/ashrae_90_1_prm/ashrae_90_1_prm_2019/ashrae_90_1_prm_2019.SpaceType.rb#L387-L392) to the `occupancy_control_credit_sum` section so the code may become more readable?

Reply:
> Also, maybe we should give an error if frac_sum would exceed 1.0.

By error you mean an error that should go to log but won't affect the execution, or the error that will halt the execution? (I think should be the second option?)
Also, that triggers another question:
What if the user data does not add up to 100%?
1. We can either raise exception or
2. We use other data to fill up the rest?

Reply:
Conclusion for exceed or under 1.0.
Both case shall be failed in a pre-processing check (new jira task)